Logging/.CSProject/LogHandler/AbstractLogHandler.cs

```diff
@@ -0,0 +1,102 @@
+        /// <summary> Log format part - the caller's source file line number </summary>
+        public const string LOG_PART_CALLER_LINE = "{5}";
+        /// <summary> Log format part - the timestamp at which the log was emitted. See <see cref="TimestampFormat"/> to customize format. </summary>
+        public const string LOG_PART_TIMESTAMP = "{6}";
+        /// <summary> Log format part - the log level. </summary>
+        public const string LOG_PART_LOG_LEVEL = "{7}";
+
+        /// <summary>
+        /// The default log format used by log handlers. Override in derived types to use a different default,
+        /// or use <see cref="LogFormat"/> to customize the format for individual log handler instances.
+        /// </summary>
+        protected virtual string DefaultLogFormat =>
+            $"({LOG_PART_CALLER_TYPE}.{LOG_PART_CALLER_METHOD}|{LOG_PART_CALLER_FILE}:{LOG_PART_CALLER_LINE}) {LOG_PART_MESSAGE}";
+
+        private string m_LogFormat;
+
+        private bool m_IncludeTimestamp = false;
+        private bool m_IncludeLogLevel = false;
+
+        /// <summary>
+        /// Defines the format of the log message.
+        /// The following wrapped in {} are substituted at runtime
+        ///  - <see cref="LOG_PART_MESSAGE"/>
+        ///  - <see cref="LOG_PART_CALLER_TYPE"/>
+        ///  - <see cref="LOG_PART_CALLER_METHOD"/>
+        ///  - <see cref="LOG_PART_CALLER_PATH"/>
+        ///  - <see cref="LOG_PART_CALLER_FILE"/>
+        ///  - <see cref="LOG_PART_CALLER_LINE"/>
+        ///  - <see cref="LOG_PART_TIMESTAMP"/>
+        ///  - <see cref="LOG_PART_LOG_LEVEL"/>
+        ///
+        /// Default: "({LOG_PART_CALLER_TYPE}.{LOG_PART_CALLER_METHOD}|{LOG_PART_CALLER_FILE}:{LOG_PART_CALLER_LINE}) {LOG_PART_MESSAGE}"
+        /// </summary>
+        /// <example>
+        /// The default format produces "(MyType.MyMethod|MyFile:12) Message" for a log "Message" issued in the
+        /// file "MyFile" on line "12" from type "MyType" and method "MyMethod".
+        /// </example>
+        public string LogFormat
+        {
+            get { return m_LogFormat; }
+            set
+            {
+                m_LogFormat = value;
+
+                m_IncludeLogLevel = m_LogFormat.Contains(LOG_PART_LOG_LEVEL);
+                m_IncludeTimestamp = m_LogFormat.Contains(LOG_PART_TIMESTAMP);
+            }
+        }
+
+        /// <summary>
+        /// Defines the format of the timestamp written to the log, when included in the log format
+        /// Default: HH:mm:ss (ex: 12:34:56)
+        /// </summary>
+        public string TimestampFormat { get; set; } = "HH:mm:ss";
+
+        protected AbstractLogHandler()
+        {
+            LogFormat = DefaultLogFormat;
+        }
+
+        internal void HandleLog(LogLevel level, string message, in CallerInfo callerInfo)
+        {
+            HandleFormattedLog(level, string.Format(
+                m_LogFormat,
+                message,
```

Comment:
I marked this (and `CallerInfo`) `internal` as they are entirely contained within the logging assembly. If keeping this public is preferred, either so `myLogHandler.HandleLog(...)` can be called directly or so `callerInfo` can be forwarded to `HandleFormattedLog` too, let me know and I can change it, but it felt cleaner to keep the log formatting details wrapped up within Anvil, and simply pass a `LogLevel` and formatted string to handlers.

Reply:
I think that makes sense to me

Reply:
In its current form this is fine but depends on the outcome of this comment https://github.com/decline-cookies/anvil-csharp-core/pull/123#discussion_r997490364